Modernize shouldBeEmpty and shouldNotBeEmpty collection assertions

okarmazin opened the pull request:
- Add assertions for all primitive arrays (#4354)
- Matcher now prints a more precise collection type.
- shouldBeEmpty failures now print size of the failed collection, if size is well known.
- Out-of-place Sequence tests were moved to the appropriate file. While they were testing the same operation (isEmpty), Sequence matchers live in their own separate package. The tests were relocated to that same package.
- Iterable.shouldBeEmpty no longer deadlocks on infinite iterables.

okarmazin committed:
Prepare parameterized matchers

Will allow reusing the matchers without code duplication

okarmazin committed:
Use the parameterized matcher factories for isEmpty assertions

okarmazin committed:
Replace "should beEmpty()" with ".shouldBeEmpty()" in other tests

okarmazin committed:
Using beEmpty and beEmptyArray is now an error

okarmazin committed:
Add test cases for isNotEmpty

okarmazin committed:
Reorder functions in empty.kt for easier human visual control

Ordered by receiver, not by operation. Easier for the human eye to check.

okarmazin committed:
BooleanArray

okarmazin committed:
ByteArray

okarmazin committed:
ShortArray

okarmazin committed:
CharArray

okarmazin committed:
IntArray

okarmazin committed:
LongArray

okarmazin committed:
FloatArray

okarmazin committed:
DoubleArray

okarmazin committed:
Report actual collection size in isEmpty failures

AlexCue987 commented on kotest-property/src/jvmTest/kotlin/com/sksamuel/kotest/property/shrinking/FloatShrinkerTest.kt:
```diff
@@ -21,7 +20,7 @@ class FloatShrinkerTest : FunSpec() {
 
          forAll(*values) { value ->
-            FloatShrinker.shrink(value) should beEmpty()
+            FloatShrinker.shrink(value).shouldBeEmpty()
```
`should beEmpty()` and `shouldNot beEmpty()` should be supported, it's part of commonly used style, and most assertions do support both `shouldXxx()` and `should Xxx()`. Is there a reason this PR makes `beEmpty` private? I'd advocate for reverting that change, so that this test, and presumably many others in other code bases, can be kept as is.

okarmazin replied:
Okay. In that case I'll revert this as unrelated I'll add `beEmpty` matcher factories for all the newly added arrays so that matcher-style and assertion-style are on par.

AlexCue987 reviewed (changes requested):
Overall a very nice addition of features, but should revert some breaking changes.

okarmazin committed:
Colocate Sequence isEmpty matchers with other sequence matchers

While it is the same assertion, sequence matchers are separated from collection matchers. These tests were out of place.

okarmazin committed:
Undeprecate beEmpty and beEmptyArray

okarmazin committed:
Generalize isEmpty matcher from Collection to Iterable

okarmazin committed:
Protect Iterable.shouldBeEmpty against long/infinite iterables

okarmazin committed:
Minor cleanup

okarmazin committed:
Double/FloatShrinker Revert to "should beEmpty"

AlexCue987 commented on kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/empty.kt:
```diff
@@ -7,50 +7,140 @@ import io.kotest.matchers.invokeMatcher
+   if (this == null) fail("FloatArray")
+   asList() should beEmpty("FloatArray")
+   return this
+}
```
these functions look very similar. have you explored the possibility of implementing the logic once via a generic with reified type?

okarmazin replied:
I don't think there is any reasonable way to change this since `FloatArray != Array<Float>`. There is no common supertype for primitive arrays. If you wanted a single generic function `shouldBeEmpty` that works across all primitive arrays, you would have to make it an extension on any `T` - and that is not desirable IMO.

okarmazin replied:
Right now I believe a reasonable balance has been struck - all significant logic is extracted to a single matcher and the assertions which use it are organized in a way that can easily be checked by human eye for correctness.

okarmazin committed:
Merge branch 'master' into okarmazin/collection-assertions-primitive-empty
